File: cpp/se/AuDomainModifyRegistrantCommand.cpp
#include "se/AuDomainModifyRegistrantCommand.hpp"
#include "common/ErrorPkg.hpp"
#include "se/AuExtension.hpp"
#include "xml/XMLHelper.hpp"

namespace {
    AuExtension& auExtension() {
        static AuExtension* auExtension = new AuExtension();
        return *auExtension;
    }
}; // anonymous namespace

AuDomainModifyRegistrantCommand::AuDomainModifyRegistrantCommand(
        const std::string& name,
        const std::string& registrantName,
        const std::string& explanation,
        const std::string* eligibilityType,
        int policyReason,
        const std::string* registrantID,
        const std::string* registrantIDType,
        const std::string* eligibilityName,
        const std::string* eligibilityID,
        const std::string* eligibilityIDType) : DomainUpdateCommand(name)
{
    if (eligibilityType == NULL
            || (registrantID == NULL && registrantIDType != NULL)
            || (registrantID != NULL && registrantIDType == NULL)
            || (eligibilityID == NULL && eligibilityIDType != NULL)
            || (eligibilityID != NULL && eligibilityIDType == NULL))
    {
        throw IllegalArgException(
                ErrorPkg::getMessage("se.domain.modify.au.missing_arg"));
    }

    DOMElement* auextUpdate = xmlWriter->appendChild(
            xmlWriter->appendChild(command, "extension"),
            "update",
            auExtension().getURI());

    auextUpdate->setAttribute(
            XStr("xsi:schemaLocation").str(),
            XStr(auExtension().getSchemaLocation()).str());

    DOMElement* auProperties = xmlWriter->appendChild(auextUpdate,
            "auProperties");

    XMLHelper::setTextContent(
            xmlWriter->appendChild(auProperties, "registrantName"),
            registrantName);

    if (registrantID != NULL && registrantIDType != NULL)
    {
        xmlWriter->appendChild(
                auProperties,
                "registrantID",
                *registrantID,
                "type",
                *registrantIDType);
    }

    if (eligibilityType != NULL)
    {
        XMLHelper::setTextContent(
                xmlWriter->appendChild(auProperties, "eligibilityType"),
                *eligibilityType);
    }

    if (eligibilityName != NULL)
    {
        XMLHelper::setTextContent(
                xmlWriter->appendChild(auProperties, "eligibilityName"),
                *eligibilityName);
    }

    if (eligibilityID != NULL && eligibilityIDType != NULL)
    {
        xmlWriter->appendChild(
                auProperties,
                "eligibilityID",
                *eligibilityID,
                "type",
                *eligibilityIDType);
    }

    XMLHelper::setTextContent(
            xmlWriter->appendChild(auProperties, "policyReason"),
            policyReason);

    XMLHelper::setTextContent(
            xmlWriter->appendChild(auextUpdate, "explanation"),
            explanation);
}

File: src/main/java/com/ausregistry/jtoolkit2/se/AuDomainModifyRegistrantCommand.java
package com.ausregistry.jtoolkit2.se;

import com.ausregistry.jtoolkit2.ErrorPkg;
import org.w3c.dom.Element;

/**
 * Extension of the domain mapping of the EPP update command, as defined in
 * RFC5730 and RFC5731, to .au domain names, the specification of which is in
 * the XML schema definition urn:X-au:params:xml:ns:auext-1.2.
 * This class should only be used to correct au extension data for .au domain
 * names, and only where the legal registrant has not changed.
 * Use this class to generate a standards-compliant XML document, given simple
 * input parameters.  The toXML method in Command serialises this object to
 * XML.
 */
public final class AuDomainModifyRegistrantCommand
    extends DomainUpdateCommand {

    private static final long serialVersionUID = 8196324073107340593L;

    /**
     * @throws IllegalArgumentException if {@code name}, {@code eligibilityType} or {@code registrantName}
     * is {@code null}.
     */
    public AuDomainModifyRegistrantCommand(String name, String eligibilityType,
            int policyReason, String registrantName, String registrantID,
            String registrantIDType, String eligibilityName,
            String eligibilityID, String eligibilityIDType,
            String explanation) {

        super(name);

        if (eligibilityType == null || registrantName == null) {
            throw new IllegalArgumentException(ErrorPkg.getMessage(
                    "se.domain.modify.au.missing_arg"));
        }

        assert (registrantID == null && registrantIDType == null)
            || (registrantID != null && registrantIDType != null);

        assert (eligibilityID == null && eligibilityIDType == null)
            || (eligibilityID != null && eligibilityIDType != null);

        extension = xmlWriter.appendChild(command, "extension");
        Element extensionElement = xmlWriter.appendChild(extension, "update", ExtensionImpl.AU.getURI());

        extensionElement.setAttribute(
                "xsi:schemaLocation",
                ExtensionImpl.AU.getSchemaLocation());

                Element auProperties = xmlWriter.appendChild(extensionElement, "auProperties");
        xmlWriter.appendChild(
            auProperties,
            "registrantName").setTextContent(registrantName);

        if (registrantID != null && registrantIDType != null) {
            xmlWriter.appendChild(auProperties,
                    "registrantID",
                    registrantID,
                    "type", registrantIDType);
        }

        xmlWriter.appendChild(auProperties,
            "eligibilityType").setTextContent(eligibilityType);

        if (eligibilityName != null) {
            xmlWriter.appendChild(
                                auProperties,
                    "eligibilityName").setTextContent(eligibilityName);
        }

        if (eligibilityID != null && eligibilityIDType != null) {
            xmlWriter.appendChild(auProperties,
                    "eligibilityID",
                    eligibilityID,
                    "type", eligibilityIDType);
        }

        xmlWriter.appendChild(
                        auProperties,
                "policyReason").setTextContent(
                    String.valueOf(policyReason));

        xmlWriter.appendChild(extensionElement,
                "explanation").setTextContent(explanation);
    }
}

